src/lib/y2partitioner/actions/add_bcache.rb

```diff
@@ -0,0 +1,69 @@
+# encoding: utf-8
+
+# Copyright (c) [2018] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "y2partitioner/dialogs/bcache"
+require "y2partitioner/device_graphs"
+
+Yast.import "Popup"
+
+module Y2Partitioner
+  module Actions
+    class AddBcache
+      def run
+        result = {}
+        dialog = Dialogs::Bcache.run(device_graph, result)
+        return :cancel if dialog != :next
+
+        create_device(result)
+        :finish
+      end
+
+    private
+
+      def create_device(result)
+        backing = device_graph.blk_devices.find{ |d| d.name == result[:backing] }
+        raise "Invalid result #{result.inspect}. Backing not found." unless backing
+
+        caching = device_graph.blk_devices.find{ |d| d.name == result[:caching] }
+        raise "Invalid result #{result.inspect}. Backing not found." unless backing
+
+        bcache_nums = device_graph.bcaches.map(&:basename).map { |n| n[/bcache(\d+)/, 1].to_i }
+        new_num = (bcache_nums.max || -1) + 1
+        bcache = backing.create_bcache("/dev/bcache#{new_num}")
+
+        bcache_cset = device_graph.bcache_csets.find { |d| d.blk_devices.include?(caching) }
+        if !bcache_cset
+          if caching.respond_to?(:partition_table)
```

Comment:
Why only if it responds to `#partition_table`?

Reply:
because we want to remove only partition table, nothing else. But I agree that it does not hurt if it is already there.

Reply:
Then you should not be checking whether it responds to `#partition_table` (a.k.a. if it CAN have a partition table), but whether it HAS a partition table.

Reply:
problem is that not all block devices can have it like partitions.

Reply:
but lets make it generic a just call it always.